QModManager/Utility/MainMenuMessages.cs

```diff
@@ -121,98 +137,105 @@ static IEnumerator _waitForLoad()
-                    AccessTools.Method(typeof(AddMessages), nameof(AddMessages.Postfix)));
+                Patcher.hInstance.Unpatch(AccessTools.Method(typeof(ErrorMessage), nameof(ErrorMessage.Awake)),
+                    HarmonyPatchType.Postfix, Patcher.hInstance.Id);
 
-                hInstance.Unpatch(AccessTools.Method(typeof(ErrorMessage), nameof(ErrorMessage.OnUpdate)),
-                    AccessTools.Method(typeof(Patches), nameof(UpdateMessages)));
+                Patcher.hInstance.Unpatch(AccessTools.Method(typeof(ErrorMessage), nameof(ErrorMessage.OnUpdate)),
+                    HarmonyPatchType.Transpiler, Patcher.hInstance.Id);
             }
 
-            [HarmonyPatch(typeof(ErrorMessage), nameof(ErrorMessage.Awake))]
-            internal static class AddMessages
+            [HarmonyPostfix, HarmonyPatch(typeof(ErrorMessage), nameof(ErrorMessage.Awake))]
+            private static void ErrorMessage_Awake_Postfix()
             {
-                internal static void Postfix()
-                {
-                    prevOffset = ErrorMessage.main.offset;
+                prevOffset = ErrorMessage.main.offset;
 
-                    if (!inited)
-                        return;
+                if (!inited)
+                    return;
 
-                    ErrorMessage.main.offset = newOffset;
+                ErrorMessage.main.offset = newOffset;
 
-                    messageQueue.ForEach(msg => AddInternal(msg));
-                    messageQueue.Clear();
-                }
+                messageQueue.ForEach(msg => AddInternal(msg));
+                messageQueue.Clear();
             }
 
-            private static float _getVal(float val, ErrorMessage._Message message) => messages.Contains(message) ? 1f : val;
 
             // we changing result for 'float value = Mathf.Clamp01(MathExtensions.EvaluateLine(...' to 1.0f
             // so text don't stay in the center of the screen (because of changed 'timeEnd')
-            internal static IEnumerable<CodeInstruction> UpdateMessages(IEnumerable<CodeInstruction> cins)
+            [HarmonyTranspiler, HarmonyPatch(typeof(ErrorMessage), nameof(ErrorMessage.OnUpdate))]
+            private static IEnumerable<CodeInstruction> ErrorMessage_OnUpdate_Transpiler(IEnumerable<CodeInstruction> cins)
             {
-                var list = new List<CodeInstruction>(cins);
-                int index = list.FindIndex(cin => cin.opcode == OpCodes.Stloc_S && (cin.operand as LocalBuilder)?.LocalIndex == 11);
-
-                list.InsertRange(index, new List<CodeInstruction>
+                try
                 {
-                    new CodeInstruction(OpCodes.Ldloc_S, 6),
-                    new CodeInstruction(OpCodes.Call, AccessTools.Method(typeof(Patches), nameof(_getVal)))
-                });
-
-                return list;
+                    var code = new CodeMatcher(cins).
```

Comment:
Not many of us have used `CodeMatcher` before.  
Could you add some more comments explaining how this all works?

Reply:
Hmm, I'm not sure where to put comments. Basically, it's just an advanced wrapper around searching in the list of instructions in Traverse/LINQ style. It stores current instruction index, so there is no need to create and check variables for indexes and such.
So in this case it searches for specified method from the end, stores index from the operand of the finded instruction and then continues to search from the same position and then inserts couple of instructions.
It's almost the same transpiler as before, but SN and BZ have slightly different versions of that method and local indexes are not the same for that variable.

Reply:
Just some comments explaining what these methods are doing: `MatchBack`, `Advance`, `MatchForward`, `EmitDelegate`

Reply:
Well, `CodeMatcher` have its own documentation that is accessible in Visual Studio. And as for github, isn't those method names are straightforward enough to get general idea of what is going on?

Reply:
I honestly find the fluent syntax of `CodeMatcher` to be harder to follow that a hand-crafted state machine doing the same thing.

We've gone back and forth enough here. We can continue this on Discord.